pkg/finality-grandpa/round.go

```diff
@@ -0,0 +1,691 @@
+	"golang.org/x/exp/constraints"
+)
+
+// The (voting) phases of a round, each corresponding to the type of
+// votes cast in that phase.
+type Phase uint
+
+const (
+	// The prevote phase in which [`Prevote`]s are cast.
+	PrevotePhase Phase = iota
+	// The precommit phase in which [`Precommit`]s are cast.
+	PrecommitPhase
+)
+
+type voteSignature[Vote, Signature comparable] struct {
+	Vote      Vote
+	Signature Signature
+}
+
+type Single[Vote, Signature comparable] voteSignature[Vote, Signature]
+
+type Equivocated[Vote, Signature comparable] [2]voteSignature[Vote, Signature]
+
+// The observed vote from a single voter.
+type voteMultiplicity[Vote, Signature comparable] struct {
+	value interface{}
+}
+
+// can only use type constraint interfaces as function parameters
+type voteMultiplicityValue[Vote, Signature comparable] interface {
+	Single[Vote, Signature] | Equivocated[Vote, Signature]
+}
+
+func setvoteMultiplicity[
+	Vote, Signature comparable,
+	T voteMultiplicityValue[Vote, Signature],
+](vm *voteMultiplicity[Vote, Signature], val T) {
+	vm.value = val
+}
+
+func newVoteMultiplicity[
+	Vote, Signature comparable,
+	T voteMultiplicityValue[Vote, Signature],
+](val T) (vm voteMultiplicity[Vote, Signature]) {
+	return voteMultiplicity[Vote, Signature]{
+		value: val,
+	}
+}
+
+func (vm voteMultiplicity[Vote, Signature]) Value() interface{} {
+	return vm.value
+}
+
+func (vm voteMultiplicity[Vote, Signature]) Contains(vote Vote, sig Signature) bool {
+	vs := voteSignature[Vote, Signature]{vote, sig}
+	switch in := vm.Value().(type) {
+	case Single[Vote, Signature]:
+		return voteSignature[Vote, Signature](in) == vs
+	case Equivocated[Vote, Signature]:
+		return in[0] == vs || in[1] == vs
+	default:
+		panic("should never happen")
+	}
+}
+
+type voteTracker[ID constraints.Ordered, Vote, Signature comparable] struct {
+	votes         *btree.Map[ID, voteMultiplicity[Vote, Signature]]
+	currentWeight VoteWeight
+	mtx           sync.RWMutex
+}
+
+func newVoteTracker[ID constraints.Ordered, Vote, Signature comparable]() voteTracker[ID, Vote, Signature] {
+	return voteTracker[ID, Vote, Signature]{
+		votes: btree.NewMap[ID, voteMultiplicity[Vote, Signature]](2),
+	}
+}
+
+// track a vote, returning a value containing the multiplicity of all votes from this ID
+// and a bool indicating if the vote is duplicated.
+// if the vote is the first equivocation, returns a value indicating
+// it as such (the new vote is always the last in the multiplicity).
+//
+// if the vote is a further equivocation, it is ignored and there is nothing
+// to do.
+//
+// since this struct doesn't track the round-number of votes, that must be set
+// by the caller.
+func (vt *voteTracker[ID, Vote, Signature]) AddVote(
+	id ID,
+	vote Vote,
+	signature Signature,
+	weight VoterWeight,
+) (*voteMultiplicity[Vote, Signature], bool) {
+	vt.mtx.Lock()
+	defer vt.mtx.Unlock()
+
+	var ok bool
+	vm, ok := vt.votes.Get(id)
+	if !ok {
+		// TODO: figure out saturating_add stuff
```

Comment:
still todo?

Reply:
Yea I still need to implement this.  But not of a huge importance at the moment since I don't think we'll be overflowing anytime soon.